### ISSUE TITLE
Could not specify explicit device specification ''

### ISSUE THREAD

**Issue by mrry** [tensorflow/tensorflow#1310]
Several users have reported an issue that occurs with the following steps (e.g. Issue #1297):

0. Start TensorFlow in a setting with a GPU.
1. Define some variables with no explicit `tf.device()` set. This most often happens with embedding variables (used as arguments to `tf.gather()` or `tf.embedding_lookup()`).
2. Initialize them. (They will be placed on the GPU, because it is the "best available device".)
3. On running the first training step, the following error (or similar) is raised:

```
InvalidArgumentError: Cannot assign a device to node 'Adagrad/update_Variable_2/SparseApplyAdagrad': Could not satisfy explicit device specification '' because the node was colocated with a group of nodes that required incompatible device '/job:localhost/replica:0/task:0/GPU:0'
```

The `SparseApplyAdagrad` op (or in general most `SparseApplyFoo` ops) is only defined on CPU. The variable has already been placed on GPU, so the graph is not placeable. Attempting to run the same program with `use_soft_placement=True` also fails, although with a stranger error.

**TL;DR:** If this affects you, create your variables in a `with tf.device("/cpu:0"):` block, until this issue is resolved.

The issue arises because (i) TensorFlow places variables on the first device where they run, (ii) it always prefers GPU over CPU when it is availabe, (iii) initialization ops are available on GPU, and (iv) it applies the placement algorith to the *pruned subgraph* (not the entire client graph).

One workaround would be to apply the placement algorithm to the entire client graph. (This is the approach used in the separate `master_session.cc`/`simple_graph_execution_state.cc` codepath, used in the distributed runtime.) However, this has the effect of leaving the session in a broken state as soon as an unplaceable node is encountered. Switching to this behavior might cause issues for people doing exploratory graph construction in a REPL (IPython etc.) because the only remedy would be to recreate the entire graph on seeing such an error. Therefore, while failing fast in a non-interactive setting would be fine, a different solution for interactive use might be required.

**Comment by hotpxl** [comment]
Previously I had
```python
with tf.device('/job:ps/task:0'):
    a = tf.constant(100)
    
with tf.device('/job:ps/task:1'):
    b = tf.constant(100)

with tf.Session('grpc://10.201.106.12:2222', config=tf.ConfigProto(log_device_placement=True)) as sess:
    print(sess.run(a + b))
```
and I encountered this `Could not satisfy explicit device specification '/job:ps/task:1'` error.

Then I changed to
```python
with tf.device('/job:ps/task:0'):
    with tf.device('/cpu:0'):
        a = tf.constant(100)
    
with tf.device('/job:ps/task:1'):
    with tf.device('/cpu:0'):
        b = tf.constant(100)

with tf.Session('grpc://10.201.106.12:2222', config=tf.ConfigProto(log_device_placement=True)) as sess:
    print(sess.run(a + b))
```
but I still have this error.

Am I getting the workaround right? Thanks!

**Comment by hotpxl** [comment]
Looks my problem is not related to CPU/GPU you are talking about. More likely because of not being able to find peer worker.

**Comment by hotpxl** [comment]
In my setup, I started with `./grpc_tensorflow_server --cluster_spec='worker|10.201.106.12:2222;10.201.106.11:2222' --job_name=worker --task_index=0` and the other with 0 replaced by 1. Is there any problem with the way I fire the servers? Why couldn't I place explicitly on the remote host?

Thanks

**Comment by hotpxl** [comment]
Found the problem. It should be `--task_id` not `--task_index`. Please change `readme.md` to reflect this change.

**Comment by stephenroller** [comment]
Has there been any more work to implement some of the SparseApplyOps on GPU? GPU-embeddings are dramatically faster than CPU ones, so anyone using RNNs would benefit significantly,

**Comment by vrv** [comment]
(The root bug was fixed a few weeks ago by placing the entire graph on first run, not just the pruned ones).
